### ISSUE TITLE
reproduce joint_alignment translation

### ISSUE THREAD

**Issue by hljjjmssyh** [pytorch/fairseq#2290]
### Before asking:   
1. search the issues.   
2. search the docs.    

#### What is your question?
I want to reproduce the results of the paper ( Jointly Learning to Align and Translate with Transformer Models (Garg et al., 2019).). Unfortunately I encountered some difficulties. Could you help me solve these problems?

#### Code

   
Here is the command I used.
fairseq-train \
    binarized \
    --arch transformer_align --share-all-embeddings \
    --optimizer adam --adam-betas '(0.9, 0.98)' --clip-norm 0.0 --activation-fn relu\
    --lr 0.0003 --lr-scheduler inverse_sqrt --warmup-updates 4000 --warmup-init-lr 1e-07 \
    --dropout 0.1 --attention-dropout 0.1 --weight-decay 0.0 \
    --max-tokens 3500 --label-smoothing 0.1 \
    --save-dir ./checkpoints --log-interval 1000 --max-update 60000 \
    --keep-interval-updates -1 --save-interval-updates 0 \
    --load-alignments --criterion label_smoothed_cross_entropy_with_alignment \
    --update-freq 4
I used 4 titan xp GPUs.
### Here is the output.
2020-06-30 20:52:43 | INFO | fairseq_cli.train | model transformer_align, criterion LabelSmoothedCrossEntropyCriterionWithAlignment
2020-06-30 20:52:43 | INFO | fairseq_cli.train | num. model params: 62418944 (num. trained: 62418944)
2020-06-30 20:52:43 | INFO | fairseq.trainer | detected shared parameter: encoder.embed_tokens.weight <- decoder.embed_tokens.weight
2020-06-30 20:52:43 | INFO | fairseq.trainer | detected shared parameter: encoder.embed_tokens.weight <- decoder.output_projection.weight
2020-06-30 20:52:43 | INFO | fairseq.utils | ***********************CUDA enviroments for all 4 workers***********************
2020-06-30 20:52:43 | INFO | fairseq.utils | rank   0: capabilities =  6.1  ; total memory = 11.910 GB ; name = TITAN Xp                                
2020-06-30 20:52:43 | INFO | fairseq.utils | rank   1: capabilities =  6.1  ; total memory = 11.910 GB ; name = TITAN Xp                                
2020-06-30 20:52:43 | INFO | fairseq.utils | rank   2: capabilities =  6.1  ; total memory = 11.910 GB ; name = TITAN Xp                                
2020-06-30 20:52:43 | INFO | fairseq.utils | rank   3: capabilities =  6.1  ; total memory = 11.910 GB ; name = TITAN Xp                                
2020-06-30 20:52:43 | INFO | fairseq.utils | ***********************CUDA enviroments for all 4 workers***********************
2020-06-30 20:52:43 | INFO | fairseq_cli.train | training on 4 devices (GPUs/TPUs)
2020-06-30 20:52:43 | INFO | fairseq_cli.train | max tokens per GPU = 3500 and max sentences per GPU = None
2020-06-30 20:52:43 | INFO | fairseq.trainer | no existing checkpoint found ./checkpoints/checkpoint_last.pt
2020-06-30 20:52:43 | INFO | fairseq.trainer | loading train data for epoch 1
2020-06-30 20:52:44 | INFO | fairseq.data.data_utils | loaded 5223425 examples from: binarized/train.en-de.en
2020-06-30 20:52:44 | INFO | fairseq.data.data_utils | loaded 5223425 examples from: binarized/train.en-de.de
2020-06-30 20:52:44 | INFO | fairseq.tasks.translation | binarized train en-de 5223425 examples
2020-06-30 20:52:45 | INFO | fairseq.data.data_utils | loaded 5223425 examples from: binarized/train.align.en-de
2020-06-30 21:09:51 | INFO | train_inner | epoch 001:   1000 / 3021 loss=11.477, nll_loss=10.903, alignment_loss=3.867, ppl=1914.17, wps=52579.3, ups=1.03, wpb=51073.4, bsz=1733.5, num_updates=1000, lr=7.5075e-05, gnorm=1.328, train_wall=977, wall=1027
2020-06-30 21:26:06 | INFO | train_inner | epoch 001:   2000 / 3021 loss=8.528, nll_loss=7.478, alignment_loss=2.896, ppl=178.23, wps=52335.1, ups=1.03, wpb=51056.2, bsz=1728.7, num_updates=2000, lr=0.00015005, gnorm=1.167, train_wall=974, wall=2003
2020-06-30 21:42:22 | INFO | train_inner | epoch 001:   3000 / 3021 loss=6.714, nll_loss=5.366, alignment_loss=2.326, ppl=41.24, wps=52352.6, ups=1.02, wpb=51092.4, bsz=1725.7, num_updates=3000, lr=0.000225025, gnorm=1.075, train_wall=974, wall=2979
2020-06-30 21:42:42 | INFO | fairseq.data.iterators | Data loading buffer is empty or nearly empty. This may indicate a data loading bottleneck, and increasing the number of workers (--num-workers) may help.
2020-06-30 21:42:42 | INFO | fairseq.data.iterators | Data loading buffer is empty or nearly empty. This may indicate a data loading bottleneck, and increasing the number of workers (--num-workers) may help.
2020-06-30 21:42:42 | INFO | fairseq.data.iterators | Data loading buffer is empty or nearly empty. This may indicate a data loading bottleneck, and increasing the number of workers (--num-workers) may help.
2020-06-30 21:42:42 | INFO | fairseq.data.iterators | Data loading buffer is empty or nearly empty. This may indicate a data loading bottleneck, and increasing the number of workers (--num-workers) may help.

Here are all the results of the program output. In the subsequent time, the program has no output.
@myleott

**Comment by lematt1991** [comment]
Can you try using the latest version of master?  It looks like 14ee059a36092a4216ec38601ca32be11383eecb should provide a fix.  See #2264 for context.

**Comment by hljjjmssyh** [comment]
@lematt1991 I am using the latest version, unfortunately, the problem still exists.

**Comment by myleott** [comment]
It looks like it’s hanging near the end of the epoch? How big is the validation set?

Also, please pull the latest master, we just fixed another hang in the data loader: https://github.com/pytorch/fairseq/commit/97ca0c022c74232690aae1c67f8d535602d071be
